### PR TITLE
Remove ClusterName

### PR DESCRIPTION
**Pull request by lavalamp**
Problem summary:
* We have a metadata field clusterName, added for the first take on federation many years ago.
* It’s force-cleared on read and write paths, since we couldn’t figure out what to do with it.
* Although clients will read, parse, and echo back the field, the semantics have never been defined. No one knows what this field “means”.
* It has come to my attention that at least one kube-derivative project (KCP) is using this field, by disabling the server code that clears this field.

**Main concern**: Multiple projects seeking interoperability with vanilla OSS k8s, each extending the api in similar but non-compatible ways, leading to ecosystem fracture.

Suggested way to fix this:
* As we did for CRD group names, we should reserve all unqualified field names within metadata (indeed, the entire k8s-owned API surface) for the future exclusive use of the kubernetes project.
* New policy:
  * Users wanting non-standard within-api-object extension fields therefore need to prefix any such field name with some unambiguous and unlikely-to-collide project identifier.
  * Additionally, those customizing the api in that way need to agree to transition onto a standardized k8s field within a reasonable amount of time (2 years?), in the event that k8s decides that the concept is useful and creates a standardized field for it.
* Code changes:
  * Remove the ClusterName field, to reduce the temptation to misuse it. Lack of semantics makes it both especially tempting to re-use this field, and also especially likely to be incompatible with others’ use of the field. Because this field is force cleared on read AND write, it is unusually safe to remove.

ETA: This PR renames the ClusterName field in the go struct / interface while leaving it serialized. This will help clients detect any accidental use. Next release we will remove it completely. (I personally think this is probably an excessive amount of caution, but see the discussion below)

/kind cleanup
/kind api-change
/kind deprecation

/sig arch
/sig api-machinery

#### Does this PR introduce a user-facing change?

```release-note
The metadata.clusterName field is deprecated. This field has always been unwritable and always blank, but its presence is confusing, so we will remove it next release. Out of an abundance of caution, this release we have merely changed the name in the go struct to ensure any accidental client uses are found before complete removal.
```

### CODE REVIEW

**Comment by lavalamp** [comment]
/hold

For discussion, in case anyone disagrees

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 1]
```diff
@@ -257,12 +257,6 @@ type ObjectMeta struct {
 	// +patchStrategy=merge
 	Finalizers []string `json:"finalizers,omitempty" patchStrategy:"merge" protobuf:"bytes,14,rep,name=finalizers"`
 
-	// The name of the cluster which the object belongs to.
-	// This is used to distinguish resources with same name and namespace in different clusters.
-	// This field is not set anywhere right now and apiserver is going to ignore it if set in create or update request.
-	// +optional
-	ClusterName string `json:"clusterName,omitempty" protobuf:"bytes,15,opt,name=clusterName"`
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:264]: leave a tombstone comment that proto tag 15 and json name `clusterName` were previously used and can't be re-added with an incompatible type

we've done this in the past by commenting out a field definition:
```go
	// Fields is tombstoned to show why 5 is a reserved protobuf tag.
	//Fields *Fields `json:"fields,omitempty" protobuf:"bytes,5,opt,name=fields,casttype=Fields"`
```

Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:264]: I will put it at the end of the struct to ensure no weird triggering of the "second before" comment block in the docs generator...

**Comment by deads2k** [comment]
I agree that this is a really tempting field to use, particularly client-side, but the risk of incompatible implementations and unclear semantics make it a poor choice for users in the end.

The suggestion for allowing composition using prefixed names also makes sense to me and I think it would result in a better API in the end because layers would likely be cleaner.

> As we did for CRD group names, we should reserve all unqualified field names within metadata (indeed, the entire k8s-owned API surface) for the future exclusive use of the kubernetes project.

This is a very good idea.

/lgtm
/hold for longer discussion

**Comment by smarterclayton** [comment]
> Main concern: Multiple projects seeking interoperability with vanilla OSS k8s, each extending the api in similar but non-compatible ways, leading to ecosystem fracture.

Isn't this what conformance is about?  I'm pretty sure this is why we have conformance, and how we ensure users are successful across distros.

Is sig-cluster-lifecycle using this field in client side logic?  I swear I've seen users, but haven't followed up recently.

> Remove the ClusterName field, to reduce the temptation to misuse it. Lack of semantics makes it both especially tempting to re-use this field, and also especially likely to be incompatible with others’ use of the field. Because this field is force cleared on read AND write, it is unusually safe to remove.

Would this fail validation in server side apply for anyone who has client side yaml sending it?

As someone who works with the kcp folks and advises them, and deep familiarity with the current Kube project, metadata, and how conformance works, I was a little surprised to see this PR opened and given such haste to remove a metadata field (which we've never done, at all).  It feels a bit defensive (just from my perspective), and I'd obviously like to help smooth over such defensiveness between people working in good faith in the ecosystem.

/hold

for more discussion (I can't tell whether this is a serious PR or a "start discussion PR")

**Comment by smarterclayton** [comment]
Are there any other fields that are examples of this (unused and unwritten)?  I think reclaim policy on volumes is the closest, and we discussed and made that inert a while back (although haven't come back to it recently).  I would also expect us to connect this to general deprecated fields - even when cleared and disabled, this was a GA field for clients.  Removing the field from serialization of the type doesn't break API compatibility *if* old clients sending this value post deletion in other scenarios are still successful, which is one way to categorize it.

**Comment by liggitt** [comment]
> to remove a metadata field (which we've never done, at all)

We've definitely removed unused fields from metadata. `initializers` is a good example.

**Comment by liggitt** [comment]
> Removing the field from serialization of the type doesn't break API compatibility if old clients sending this value post deletion in other scenarios are still successful, which is one way to categorize it.

Clients sending requests with the field set would not be rejected by the server, and the field would be dropped (just like sending any other unknown field).

The field is not usable within Kubernetes, and the semantics are not defined for use outside Kubernetes. This is effectively a disabled alpha field left over from the federation attempts. We wouldn't add such a field now with such weak semantics, and I don't think it is reasonable to keep it in place and attract inconsistent/incompatible use.

**Comment by smarterclayton** [comment]
> We've definitely removed unused fields which were never allowed to persist data from metadata. initializers is a good example.

Always alpha though, right?

**Comment by smarterclayton** [comment]
> This ClusterName field is unconditionally cleared in [create](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/rest/create.go#L126-L129) and [update](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/rest/update.go#L139-L142), so on unmodified kubernetes it behaves like a permanently disabled alpha feature.

If there are no places in the API that reject unknown fields, then yes.  My ignorance of whether we have those places in our public API is the clarification I'm looking for.

**Comment by liggitt** [comment]
> If there are no places in the API that reject unknown fields, then yes. My ignorance of whether we have those places in our public API is the clarification I'm looking for.

There aren't today. Once server-side field validation graduates to beta, clients that opt-into strict mode will have requests rejected if they contain unknown fields. Default server behavior will be to warn, but continue dropping unknown fields without failing.

**Comment by smarterclayton** [comment]
> Any existing use I can think of seems pretty problematic.

Sure, which has stopped approximately zero people in the history of programming.  :)  About 30m of searching didn't find anyone obviously using this on the client side, but lots and lots of close similarities.  In terms of risk I am less worried about this than many other changes, but if we do break someone subtly and cause a security issue (field dropped silently, deploy user workload from gitops to admin root cluster or something like that) I'm gonna regret it.

With a warning in place, I could probably see how we could limit that risk.  I wish we had a way to warn in the client library on serialization that this field is being used improperly that wouldn't be disruptive to clients (libraries printing warnings being extremely rude).

**Comment by liggitt** [comment]
> I wish we had a way to warn in the client library on serialization that this field is being used improperly that wouldn't be disruptive to clients (libraries printing warnings being extremely rude).

Renaming the go field to `DeprecatedClusterName` and dropping the GetClusterName()/SetClusterName() accessors in 1.24 wouldn't change serialization, but would alert client library consumers on update. That also seems like a reasonable intermediate step for a release.

**Comment by dims** [comment]
xref some CAPI related things https://github.com/kubernetes-sigs/cluster-api/pull/6306 and https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/3321

**Comment by smarterclayton** [comment]
To cover some of the novel ground from a project policy perspective:

> As we did for CRD group names, we should reserve all unqualified field names within metadata (indeed, the entire k8s-owned API surface) for the future exclusive use of the kubernetes project.

> Users wanting non-standard within-api-object extension fields therefore need to prefix any such field name with some unambiguous and unlikely-to-collide project identifier.

Unlikely to collide is human, messy, and approximate, so even if we recommend something, we should couch it in clear language of "be aware that even if you pick a good prefix, you're still at risk of being broken, and here's what will happen when it does".  I.e. describe the problem and the risk and let that guide their actions.  But of course in practice even unnamespaced fields will only collide rarely (for a given object, rate of field addition is X/year, rate of custom field is probably <<< X/year).  

Putting fields into existing objects owned by another person is only weakly solved with namespacing - the better option for that user is to declare a mapping or redefine the type under a different name, which isn't really a concept Kube has to deal with today, but which almost everyone using Kube is doing today with unstructured config manipulation of kube api objects (whether going from helm values files to fully materialized objects, or by transforming objects inline).  That leads into...

> Additionally, those customizing the api in that way need to agree to transition onto a standardized k8s field within a reasonable amount of time (2 years?), in the event that k8s decides that the concept is useful and creates a standardized field for it.

Knative duck typing and Gateway API are interesting examples of problems like this (for the latter, we realized that no amount of extensibility would fix ingress, for other reasons).  Annotations don't fit cleanly with spec/status split, which is one reason why they can't easily help with addressing extensibility.  Perhaps we've missed an opportunity to articulate "extension fields" (vs just extension metadata) in a way that kube patterns (spec, status, field manager, apply, explain, openapi) can leverage.  An extension you can't apply is effectively worthless, and then eventually you want to move it to a standard field, so perhaps finding a way to encourage an ecosystem component to explore how that would actually get easier would reduce a bit of the "shields up" perception around this tightening.

**Comment by lavalamp** [comment]
I haven't read all the discussion yet but Jordan convinced me to preserve round-tripping for an additional release. I'll update this PR when I get a chance (busy day).

**Comment by lavalamp** [comment]
> I was a little surprised to see this PR opened and given such haste to remove a metadata field

Yeah, sorry, I didn't expect you to see this PR before my heads-up message. The urgency is from the feeling that use is about to snowball and it's much easier to stop the sooner we do something. It seems like we're mostly at consensus now, so the urgency has declined :)

> If we ever decide to introduce something like a clusterName, can we reuse the field name? Why or why not?

In general, you can't reuse names, it's very bad karma. In this particular case, since the field was so thoroughly disabled, I could see making an exception, maybe. However in reality, if this concept were properly designed, I really doubt the result would be a single field -- I expect something more like a new scoping / cluster identification clause which has a UID and maybe a friendly human name. So I think it's likely academic in this case.

> If someone wants to remove another v1 stable field

IMO we will not get into this situation again in metadata, ever. I would not be in favor of removing fields that are actually usable. I would not be in favor of getting around the rule by first having a few releases that made the fields unusable.

I am in favor of removing fields that have never been usable and which we would not accept today if someone tried to add them. If there are other fields like this let me know...

> novel ground from a project policy perspective

I agree we can probably come up with something better than what I put in the OP. Part of the reason it's undefined right now is that there's no way to do it within the confines of upstream, and I didn't realize it was important to put out some guidance for forks until just now.

**Comment by lavalamp** [comment]
OK I made the discussed changes, let's see if it passes tests.

File: staging/src/k8s.io/apimachinery/pkg/api/meta/meta.go [hunk 2]
```diff
@@ -130,7 +130,7 @@ func AsPartialObjectMetadata(m metav1.Object) *metav1.PartialObjectMetadata {
 				Annotations:                m.GetAnnotations(),
 				OwnerReferences:            m.GetOwnerReferences(),
 				Finalizers:                 m.GetFinalizers(),
-				ClusterName:                m.GetClusterName(),
+				ZZZ_DeprecatedClusterName:                m.GetZZZ_DeprecatedClusterName(),
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/meta/meta.go:133]: needs gofmt?

File: staging/src/k8s.io/apiserver/pkg/registry/rest/create.go [hunk 3]
```diff
@@ -123,9 +123,9 @@ func BeforeCreate(strategy RESTCreateStrategy, ctx context.Context, obj runtime.
 		objectMeta.SetManagedFields(nil)
 	}
 
-	// ClusterName is ignored and should not be saved
-	if len(objectMeta.GetClusterName()) > 0 {
-		objectMeta.SetClusterName("")
+	// ZZZ_DeprecatedClusterName is ignored and should not be saved
+	if len(objectMeta.GetZZZ_DeprecatedClusterName()) > 0 {
+		objectMeta.SetZZZ_DeprecatedClusterName("")
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/create.go:128]: issue a warning to the requester here, e.g.:

```go
warning.AddWarning(ctx, "", "metadata.clusterName was specified, but is not preserved and the API field will be removed in 1.25")
```

File: staging/src/k8s.io/apiserver/pkg/registry/rest/update.go [hunk 4]
```diff
@@ -136,9 +136,9 @@ func BeforeUpdate(strategy RESTUpdateStrategy, ctx context.Context, obj, old run
 
 	strategy.PrepareForUpdate(ctx, obj, old)
 
-	// ClusterName is ignored and should not be saved
-	if len(objectMeta.GetClusterName()) > 0 {
-		objectMeta.SetClusterName("")
+	// ZZZ_DeprecatedClusterName is ignored and should not be saved
+	if len(objectMeta.GetZZZ_DeprecatedClusterName()) > 0 {
+		objectMeta.SetZZZ_DeprecatedClusterName("")
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/update.go:141]: issue a warning to the requester here, e.g.:

```go
warning.AddWarning(ctx, "", "metadata.clusterName was specified, but is not preserved and the API field will be removed in 1.25")
```

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 5]
```diff
@@ -257,11 +257,11 @@ type ObjectMeta struct {
...
+	// ClusterName tombstone. It consumed proto tag 15. We will remove
+	// completely sans a comment here next release (1.25). The name is
+	// changed here for a release to ensure any users (which we do not
+	// expect) will get a compile error before a serialization error.
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:263]: keep the `+optional` comment, and add `Deprecated: ...` to trigger govet warnings on use as well

File: staging/src/k8s.io/apimachinery/pkg/api/meta/meta.go (hunk 2 above)
Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/api/meta/meta.go:133]: That's what I sed!

File: staging/src/k8s.io/apiserver/pkg/registry/rest/create.go (hunk 3 above)
Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/registry/rest/create.go:128]: done

File: staging/src/k8s.io/apiserver/pkg/registry/rest/update.go (hunk 4 above)
Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/registry/rest/update.go:141]: done

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go (hunk 5 above)
Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: done

**Comment by lavalamp** [comment]
Looks like tests are passing. Do we need more discussion on this or are we good to go?

**Comment by lavalamp** [comment]
OK, I finally stopped being duped by `make generated_files` and actually ran all the right scripts.

This is reeeeeally ugly, are we SURE we have to inflict this on the client libraries etc?

**Comment by dims** [comment]
> OK, I finally stopped being duped by `make generated_files` and actually ran all the right scripts.
> 
> This is reeeeeally ugly, are we SURE we have to inflict this on the client libraries etc?

Please let us not. This is just too much 😩 😭

File: staging/src/k8s.io/client-go/applyconfigurations/admissionregistration/v1/mutatingwebhookconfiguration.go [hunk 6]
```diff
@@ -232,12 +232,12 @@ func (b *MutatingWebhookConfigurationApplyConfiguration) WithFinalizers(values .
 	return b
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/applyconfigurations/admissionregistration/v1/mutatingwebhookconfiguration.go:232]: I'd suggest blocking this field here to drop it from apply configurations now:

```
func blocklisted(t *types.Type, member types.Member) bool {
	if objectMeta.Name == t.Name && member.Name == "ManagedFields" {
		return true
	}
	if objectMeta.Name == t.Name && member.Name == "SelfLink" {
		return true
	}
```

Comment (lavalamp) [reply staging/src/k8s.io/client-go/applyconfigurations/admissionregistration/v1/mutatingwebhookconfiguration.go:232]: I didn't realize it'd be that easy.

**Comment by dims** [comment]
> OK, now we're net negative LoC, that's much better.

Nice!!!

**Comment by lavalamp** [comment]
Judging by the tests, it is not, in fact, that easy

**Comment by liggitt** [comment]
> Judging by the tests, it is not, in fact, that easy

need to clear the field in the applyconfigurations fuzzing, just like managed fields:

```
			func(s *v1.ObjectMeta, c fuzz.Continue) {
				c.FuzzNoCustom(s)
				s.ManagedFields = nil
			},
```

**Comment by lavalamp** [comment]
23948rd time is the charm...

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/meta.go [hunk 7]
```diff
@@ -59,8 +59,8 @@ type Object interface {
 	SetFinalizers(finalizers []string)
 	GetOwnerReferences() []OwnerReference
 	SetOwnerReferences([]OwnerReference)
-	GetClusterName() string
-	SetClusterName(clusterName string)
+	GetZZZ_DeprecatedClusterName() string
+	SetZZZ_DeprecatedClusterName(clusterName string)
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/meta.go:63]: I'm trying to decide if the interface change is better or worse than marking the interface methods deprecated (so linters would flag) and then dropping the methods in 1.25.

The struct field change, godoc, warning, and applyconfigurations changes were all much lower churn and those lgtm.

Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/meta.go:63]: The point is to give clients compile errors, and it turns out that most code (at least in k/k) interacts with metadata via this interface, so not changing it here would at least partly defeat the purpose.

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/meta.go:63]: fair. the current state seems reasonable to me. will give a chance for others to chime in

**Comment by lavalamp** [comment]
Is pull-kubernetes-node-e2e-containerd failing for everyone or did I somehow change the time encoding?

**Comment by liggitt** [comment]
it's flaky

**Comment by lavalamp** [comment]
Ah time to change my mind on the reliability KEP then ;)

**Comment by dims** [comment]
> Ah time to change my mind on the reliability KEP then ;)

ROTFL!!!

**Comment by aojea** [comment]
> Ah time to change my mind on the reliability KEP then ;)

/honk

**Comment by lavalamp** [comment]
Any objections to getting this merged today?

**Comment by dims** [comment]
> Any objections to getting this merged today?

I see one hold from you Daniel and another from @deads2k so if both of you are ok. I don't have a objection. The only other thing i can think of is do we need to update guidance for deprecation or something based on this problem/example (but that can happen after this lands)

**Comment by lavalamp** [comment]
Consider my hold canceled, I think this was enough discussion. I'm pretty sure David is OK too, but I pinged him.

**Comment by deads2k** [comment]
I think we reached consensus.

/hold cancel

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/meta.go (hunk 7 above)
Comment (deads2k) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/meta.go:63]: I'm ok with the interface change.

**Comment by lavalamp** [comment]
/retest

...seems go1.18 has made some tests flaky...

**Pull request merged**